File: app/src/main/java/org/puffinbasic/file/PuffinBasicFile.java
package org.puffinbasic.file;

import org.jetbrains.annotations.Nullable;
import org.puffinbasic.domain.PuffinBasicSymbolTable;

import it.unimi.dsi.fastutil.ints.IntList;

public interface PuffinBasicFile {

    int DEFAULT_RECORD_LEN = 128;

    void setFieldParams(
            PuffinBasicSymbolTable symbolTable,
            IntList recordParts
    );

    int getCurrentRecordNumber();

    long getFileSizeInBytes();

    String readLine();

    byte[] readBytes(int n);

    void print(String s);

    void writeByte(byte b);

    boolean eof();

    void put(@Nullable Integer recordNumber, PuffinBasicSymbolTable symbolTable);

    void get(@Nullable Integer recordNumber, PuffinBasicSymbolTable symbolTable);

    boolean isOpen();

    void close();

    enum FileOpenMode {
        INPUT,
        OUTPUT,
        APPEND,
        RANDOM
    }

    enum FileAccessMode {
        READ_ONLY("r"),
        WRITE_ONLY("w"),
        READ_WRITE("rw");

        public final String mode;

        FileAccessMode(String mode) {
            this.mode = mode;
        }
    }

    enum LockMode {
        SHARED,
        READ,
        WRITE,
        READ_WRITE,
        DEFAULT
    }

    enum FileState {
        OPEN,
        CLOSED
    }
}

File: app/src/main/java/org/puffinbasic/file/PuffinBasicFile.kt
package org.puffinbasic.file

import org.puffinbasic.domain.PuffinBasicSymbolTable

interface PuffinBasicFile {
    fun setFieldParams(
        symbolTable: PuffinBasicSymbolTable,
        recordParts: List<Int>
    )

    val currentRecordNumber: Int
    val fileSizeInBytes: Long
    fun readLine(): String?
    fun readBytes(n: Int): ByteArray?
    fun print(s: String)
    fun writeByte(b: Byte)
    fun eof(): Boolean
    fun put(recordNumber: Int, symbolTable: PuffinBasicSymbolTable)
    operator fun get(recordNumber: Int, symbolTable: PuffinBasicSymbolTable)
    fun isOpen(): Boolean
    fun close()
    enum class FileOpenMode {
        INPUT, OUTPUT, APPEND, RANDOM
    }

    enum class FileAccessMode(@JvmField val mode: String) {
        READ_ONLY("r"), WRITE_ONLY("w"), READ_WRITE("rw")
    }

    enum class LockMode {
        SHARED, READ, WRITE, READ_WRITE, DEFAULT
    }

    enum class FileState {
        OPEN, CLOSED
    }

    companion object {
        const val DEFAULT_RECORD_LEN = 128
    }
}

interface PuffinBasicExtendedFile : PuffinBasicFile {
    fun inputDialog(prompt: String): String?
}
